contents/blog/mobile-session-replay.md

```diff
@@ -0,0 +1,139 @@
+---
+title: How we built mobile replay (and why it took so long)
+date: 2024-10-08
+author:
+ - ian-vanagas
+ - manoel-aranda-neto
+rootpage: /blog
+featuredImage: >- 
+  https://res.cloudinary.com/dmukukwp6/image/upload/posthog.com/contents/images/blog/posthog-engineering-blog.png
+featuredImageType: full
+tags:
+ - Guides
+---
+
+[Session replay](/session-replay) is one of the most powerful tools for understanding user behavior. Web session replay has been a core part of PostHog for a long time now (it was built in our first hackathon), but mobile teams have had to wait longer.
+
+Finally, it's here on [iOS](/docs/session-replay/ios), [Android](/docs/session-replay/android), and [React Native](/docs/session-replay/react-native) (with [Flutter](https://github.com/PostHog/posthog-flutter/issues/69) coming soon). 
+
+What took so long? Recording them in mobile apps is much trickier than in web apps. This post goes over why and how we finally managed to make it happen.
+
+## What's so difficult about mobile session replay?
+
+Developers have complained about [the lack of good mobile replay options](https://medium.com/goodones/15-years-later-there-is-still-no-good-session-replay-for-ios-f8d335999737). Why is that the case?
+
+### 1. Multiple platforms
+
+The industry's big secret about web session replay is that it largely relies on a single open source library to work: [rrweb](https://github.com/rrweb-io/rrweb). It includes tools for recording web interactions and state changes, structuring session data, and playback. 
+
+Unfortunately, rrweb for mobile doesn't exist. To build mobile session replay, we needed to do all the work ourselves, and this was a lot. Instead of a single JavaScript SDK, mobile requires multiple (like iOS, Android, React Native, and Flutter). 
+
+There are even breaking differences within platforms. For example, Jetpack Compose uses a compositional model for UI, which is different from Android's traditional view-based model. This means you need to develop separate ways of doing replays when using it. iOS has a similar problem with SwiftUI versus UIKit.
+
+### 2. Performance
+
+Phones are much less powerful than desktops. Because of this, we need to be much more sensitive about performance. 
+
+If you've ever tried to record your phone screen, you would know its impact on performance. Apps take longer to load, animations become choppy, reactivity degrades, and your phone heats up.
+
+Anything that degrades your user experience is not an option for many developers.
```

Comment:
This could do with a bit more caveating and context. Something about how the vast majority of phones aren't top-of-the-range iPhones. Your app needs to work just as well on $150 Android phone as a $1,500 iPhone 16 etc. That screen recording example is not universally true, so probably not a great example.